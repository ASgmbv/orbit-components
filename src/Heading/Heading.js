// @flow
import * as React from "react";
import styled from "styled-components";
import defaultTokens from "@kiwicom/orbit-design-tokens";
import createComponentFromTagProp from "react-create-component-from-tag-prop";

export const ELEMENT_OPTIONS = {
  H1: "h1",
  H2: "h2",
  H3: "h3",
  H4: "h4",
  H5: "h5",
};

export const TYPE_OPTIONS = {
  DISPLAY: "display",
  TITLE1: "title1",
  TITLE2: "title2",
  TITLE3: "title3",
};

type Props = {|
  element: $Values<typeof ELEMENT_OPTIONS>,
  type: $Values<typeof TYPE_OPTIONS>,
  children: React.Node,
  theme: typeof defaultTokens,
|};

const HeadingElement = createComponentFromTagProp({
  prop: "element",
  propsToOmit: ["type", "tokens", "theme"],
});

const StyledHeading = styled(HeadingElement)`
  font-family: ${props => props.theme.fontFamily};
  font-size: ${props => props.tokens.sizeHeading[props.type]};
  font-weight: ${props => props.tokens.weightHeading[props.type]};
  color: ${props => props.theme.colorHeading};
  line-height: ${props => props.theme.lineHeightHeading};
  margin: 0;
`;

const Heading = (props: Props) => {
  const { children, theme, type, element } = props;
  const tokens = {
    weightHeading: {
      [TYPE_OPTIONS.DISPLAY]: theme.fontWeightHeadingDisplay,
      [TYPE_OPTIONS.TITLE1]: theme.fontWeightHeadingLevel1,
      [TYPE_OPTIONS.TITLE2]: theme.fontWeightHeadingLevel2,
      [TYPE_OPTIONS.TITLE3]: theme.fontWeightHeadingLevel3,
    },
    sizeHeading: {
      [TYPE_OPTIONS.DISPLAY]: theme.fontSizeHeadingDisplay,
      [TYPE_OPTIONS.TITLE1]: theme.fontSizeHeadingTitle1,
      [TYPE_OPTIONS.TITLE2]: theme.fontSizeHeadingTitle2,
      [TYPE_OPTIONS.TITLE3]: theme.fontSizeHeadingTitle3,
    },
  };
  return (
    <StyledHeading theme={theme} tokens={tokens} type={type} element={element}>
      {children}
    </StyledHeading>
  );
};

Heading.defaultProps = {
  element: "h1",
  type: "title1",
  theme: defaultTokens,
};

export default Heading;
