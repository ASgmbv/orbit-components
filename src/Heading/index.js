// @flow
import * as React from "react";
import * as tokens from "@kiwicom/orbit-design-tokens";

const weightHeading = {
  display: tokens.fontWeightHeadingDisplay,
  title1: tokens.fontWeightHeadingLevel1,
  title2: tokens.fontWeightHeadingLevel2,
  title3: tokens.fontWeightHeadingLevel3,
};

const sizeHeading = {
  display: tokens.fontSizeHeadingDisplay,
  title1: tokens.fontSizeHeadingTitle1,
  title2: tokens.fontSizeHeadingTitle2,
  title3: tokens.fontSizeHeadingTitle3,
};

type Props = {
  element: "h1" | "h2" | "h3" | "h4" | "h5",
  type: $Keys<typeof weightHeading>,
  children: React.Node,
};

function resolveScopedStyles(scope) {
  return {
    className: scope.props.className,
    styles: scope.props.children,
  };
}

const Heading = (props: Props) => {
  const { type, element, children } = props;

  const scoped = resolveScopedStyles(
    <scope>
      <style jsx>
        {`
           {
            font-family: ${tokens.fontFamily};
            font-size: ${sizeHeading[type]};
            font-weight: ${weightHeading[type]};
            color: ${tokens.colorHeading};
            line-height: ${tokens.lineHeightHeading};
            margin: 0;
          }
        `}
      </style>
    </scope>,
  );

  return (
    <React.Fragment>
      {React.createElement(
        element,
        {
          className: [scoped.className].join(" "),
        },
        children,
      )}
      {scoped.styles}
    </React.Fragment>
  );
};

Heading.defaultProps = {
  element: "h1",
  type: "title2",
};

export default Heading;
