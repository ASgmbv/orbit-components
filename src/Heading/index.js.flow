// @flow
import type { ReactComponentStyled } from "styled-components";

import type { Globals } from "../common/common.js.flow";

type Element = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
type Type = "display" | "title1" | "title2" | "title3";

export type Props = {|
  +element?: Element,
  +type?: Type,
  +children: React$Node,
  +inverted?: boolean,
  ...Globals,
|};

declare export default React$ComponentType<Props>;

declare export var StyledHeading: ReactComponentStyled<>;
