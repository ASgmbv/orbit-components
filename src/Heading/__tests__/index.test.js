// @flow
import * as React from "react";
import { shallow } from "enzyme";

import Heading from "../index";

const children = "Children title";

describe("Heading with defaultProps", () => {
  const component = shallow(<Heading>{children}</Heading>);
  it("should contain children", () => {
    expect(
      component
        .find("Heading__StyledHeading")
        .render()
        .text(),
    ).toBe(children);
  });
  it("should match snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
