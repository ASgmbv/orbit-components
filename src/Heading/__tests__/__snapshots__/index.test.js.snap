// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendered with children Should contain title 1`] = `
<React.Fragment>
  <h1
    className="jsx-2937774018"
  >
    Children title
  </h1>
  <JSXStyle
    css=".__jsx-style-dynamic-selector{font-family:\\"Roboto\\", -apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif;font-size:22px;font-weight:500;color:#171b1e;line-height:1.2;margin:0;}"
    dynamic={
      Array [
        "\\"Roboto\\", -apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif",
        "22px",
        "500",
        "#171b1e",
        "1.2",
      ]
    }
    styleId="719312688"
  />
</React.Fragment>
`;
